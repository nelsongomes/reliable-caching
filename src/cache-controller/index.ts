export * from "./controller-interface";
export * from "./redis-cache-controller";
