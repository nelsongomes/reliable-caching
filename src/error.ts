import createHttpError from "http-errors";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
export function normalizeError(error: any): Error {
  if (error instanceof Error) {
    return error;
  } else if (typeof error === "string") {
    try {
      const parsedError = JSON.parse(error);

      switch (parsedError.type) {
        case "Error":
          return Object.assign(new Error(), parsedError);
        case "EvalError":
          return Object.assign(new EvalError(), parsedError);
        case "InternalServerError":
        case "BadRequestError":
        case "NotFoundError":
          return Object.assign(
            createHttpError(
              parsedError.statusCode || 500,
              parsedError.message || "Unknown error"
            ),
            parsedError
          );
        case "RangeError":
          return Object.assign(new RangeError(), parsedError);
        case "ReferenceError":
          return Object.assign(new ReferenceError(), parsedError);
        case "SyntaxError":
          return Object.assign(new SyntaxError(), parsedError);
        case "TypeError":
          return Object.assign(new TypeError(), parsedError);
        case "URIError":
          return Object.assign(new URIError(), parsedError);
      }
    } catch (e) {
      // failed to parse JSON
      return new Error(error);
    }
  }

  return new Error(error);
}
