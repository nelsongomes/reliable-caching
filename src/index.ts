export * from "./metrics";
export * from "./keys";
export * from "./sign";
export * from "./cache-controller";
export * from "../test/utils";
