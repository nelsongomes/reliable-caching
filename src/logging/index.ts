export enum LogLevel {
  Debug = "debug",
  Error = "error",
  Info = "info",
}

export type logHandle = (
  logLevel: LogLevel,
  message: string,
  error?: unknown
) => void;
