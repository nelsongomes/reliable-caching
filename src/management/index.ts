export * from "./generic-manager";
export * from "./management-interface";
