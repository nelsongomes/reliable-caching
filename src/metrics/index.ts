export * from "./cache-stats";
export * from "./combine-stats";
