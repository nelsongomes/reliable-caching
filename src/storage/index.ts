export * from "./redis";
export * from "./lru-in-memory";

export const SIGNATURE_SEPARATOR = ":";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type StorageWrapper<T = any> = {
  value: T;
};

// eslint-disable-next-line @typescript-eslint/ban-types
export function deepFreeze(obj: Object): Object {
  for (const key of Object.keys(obj)) {
    const parameter = (obj as never)[key];

    if (typeof parameter === "object") {
      deepFreeze(parameter);
    }
  }

  Object.freeze(obj);

  return obj;
}
