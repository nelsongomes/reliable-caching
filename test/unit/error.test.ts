import { err as serializer } from "pino-std-serializers";
import createHttpError, { HttpError } from "http-errors";
import { normalizeError } from "../../src/error";

describe("Error serialization/deserialization tests", () => {
  it("Should serialize/deserialize generic Error", async () => {
    const serialized = JSON.stringify(serializer(new Error("some error")));
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(Error);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize unknown serialized content", async () => {
    const serialized = JSON.stringify({ anyobj: "abc" });
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(Error);
  });

  it("Should serialize/deserialize unknown unserializable content", async () => {
    const serialized = "some trash";
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(Error);
  });

  it("Should serialize/deserialize TypeError", async () => {
    const serialized = JSON.stringify(serializer(new TypeError("some error")));
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(TypeError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize URIError", async () => {
    const serialized = JSON.stringify(serializer(new URIError("some error")));
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(URIError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize SyntaxError", async () => {
    const serialized = JSON.stringify(
      serializer(new SyntaxError("some error"))
    );
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(SyntaxError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize ReferenceError", async () => {
    const serialized = JSON.stringify(
      serializer(new ReferenceError("some error"))
    );
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(ReferenceError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize RangeError", async () => {
    const serialized = JSON.stringify(serializer(new RangeError("some error")));
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(RangeError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize EvalError", async () => {
    const serialized = JSON.stringify(serializer(new EvalError("some error")));
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(EvalError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize HttpError 500 error", async () => {
    const serialized = JSON.stringify(serializer(createHttpError()));
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(HttpError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize HttpError 400 error", async () => {
    const serialized = JSON.stringify(
      serializer(createHttpError(400, "bad request"))
    );
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(HttpError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize HttpError 404 error", async () => {
    const serialized = JSON.stringify(
      serializer(createHttpError(404, "not found"))
    );
    const unserialized = normalizeError(serialized);

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(HttpError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toBe(serialized);
  });

  it("Should serialize/deserialize http error with no defaults", async () => {
    const unserialized = normalizeError(
      JSON.stringify({
        type: "InternalServerError",
        stack: "...",
        status: 500,
        expose: false,
      })
    );

    expect(typeof unserialized).toBe("object");
    expect(unserialized).toBeInstanceOf(HttpError);

    // new error hydrated from JSON has the same serialized content
    expect(JSON.stringify(serializer(unserialized))).toMatchSnapshot();
  });
});
