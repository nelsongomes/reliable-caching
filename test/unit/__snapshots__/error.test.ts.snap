// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error serialization/deserialization tests Should serialize/deserialize http error with no defaults 1`] = `"{"type":"InternalServerError","message":"Unknown error","stack":"...","status":500,"expose":false,"statusCode":500}"`;
